background/index.ts

```diff
@@ -94,6 +95,126 @@ async function main() {
     { pendingTransaction: {} }
   )
 
+  // Hold the current provider and block subscriber
+  const indexBalanceContext: {
+    provider?: JsonRpcProvider
+    blockSubscriber?: (blockNumber: number) => Promise<void>
+  } = {}
+
+  const indexBalanceOnBlock = async () => {
+    const { network, networkActive } = storage.get()
+    const { nodeRpcUrl, chainId } = network[networkActive]
+
+    // Set `{ staticNetwork: true }` to avoid infinite retries if nodeRpcUrl fails.
+    // Refer: https://github.com/ethers-io/ethers.js/issues/4377
+    indexBalanceContext.provider = new JsonRpcProvider(nodeRpcUrl, chainId, {
+      staticNetwork: true
+    })
+
+    indexBalanceContext.blockSubscriber = async (blockNumber: number) => {
+      const { network, networkActive } = storage.get()
+      const { nodeRpcUrl, chainId, accountActive } = network[networkActive]
+
+      // Update provider and subscriber on network switch
+      if (
+        (await indexBalanceContext.provider.getNetwork()).chainId !==
+        number.toBigInt(chainId)
+      ) {
+        indexBalanceContext.provider.off(
+          "block",
+          indexBalanceContext.blockSubscriber
+        )
+
+        indexBalanceContext.provider = new JsonRpcProvider(
+          nodeRpcUrl,
+          chainId,
+          {
+            staticNetwork: true
+          }
+        )
+
+        indexBalanceContext.provider.on(
+          "block",
+          indexBalanceContext.blockSubscriber
+        )
+        return
+      }
```

Comment:
and then call indexBalanceOnBlock() here if we found network is changed.
```suggestion
      if (
        (await indexBalanceContext.provider.getNetwork()).chainId !==
        number.toBigInt(chainId)
      ) {
        indexBalanceOnBlock()
        return
      }
```